shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java

```diff
@@ -0,0 +1,98 @@
+    private static final String CLIENT_ID = "client_id";
+    protected final ConfigurationService configurationService;
+    protected final SessionService sessionService;
+    protected final ClientSessionService clientSessionService;
+
+    protected BaseOrchestrationHandler(
+            ConfigurationService configurationService,
+            SessionService sessionService,
+            ClientSessionService clientSessionService) {
+        this.configurationService = configurationService;
+        this.sessionService = sessionService;
+        this.clientSessionService = clientSessionService;
+    }
+
+    protected BaseOrchestrationHandler(ConfigurationService configurationService) {
+        this.configurationService = configurationService;
+        this.sessionService = new SessionService(configurationService);
+        this.clientSessionService = new ClientSessionService(configurationService);
+    }
+
+    @Override
+    public APIGatewayProxyResponseEvent handleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        return segmentedFunctionCall(
+                "ipv-api::" + getClass().getSimpleName(),
+                () -> validateAndHandleRequest(input, context));
+    }
+
+    public abstract APIGatewayProxyResponseEvent handleRequestWithUserSession(
+            APIGatewayProxyRequestEvent input, Context context, final UserSession userSession);
+
+    private APIGatewayProxyResponseEvent validateAndHandleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        ThreadContext.clearMap();
+
+        var session = sessionService.getSessionFromRequestHeaders(input.getHeaders()).orElse(null);
+        if (Objects.isNull(session)) {
+            LOG.warn("Session cannot be found");
+            return generateApiGatewayProxyErrorResponse(400, ErrorResponse.ERROR_1000);
+        } else {
+            attachSessionIdToLogs(session);
+        }
+
+        UserSession.Builder userSessionBuilder = UserSession.builder(session);
+
+        String clientSessionId =
+                getHeaderValueFromHeaders(
+                        input.getHeaders(),
+                        CLIENT_SESSION_ID_HEADER,
+                        configurationService.getHeadersCaseInsensitive());
+        userSessionBuilder.withClientSessionId(clientSessionId);
+
+        Optional<ClientSession> clientSession =
+                clientSessionService.getClientSessionFromRequestHeaders(input.getHeaders());
+        userSessionBuilder.withClientSession(clientSession.orElse(null));
+
+        var clientID =
+                clientSession
+                        .map(ClientSession::getAuthRequestParams)
+                        .map(t -> t.get(CLIENT_ID))
+                        .flatMap(v -> v.stream().findFirst());
+        attachLogFieldToLogs(LogLineHelper.LogFieldName.CLIENT_ID, clientID.orElse(UNKNOWN));
+        userSessionBuilder.withClientId(clientID.orElse(null));
+
```

Comment:
Would be good to have the persistent session ID stuff here too, but not fail if it's not present

Reply:
I think elsewhere with the persistent session we just add to logs - is that what you mean?